src/expr.rs

```diff
@@ -633,15 +658,22 @@ fn rewrite_closure(
         }
 
         // Either we require a block, or tried without and failed.
-        return rewrite_closure_block(&block, prefix, context, body_shape);
-    }
-
-    if let Some(rw) = rewrite_closure_expr(body, &prefix, context, body_shape) {
-        return Some(rw);
+        rewrite_closure_block(&block, &prefix, context, body_shape)
+    } else {
+        rewrite_closure_expr(body, &prefix, context, body_shape).or_else(|| {
+            // The closure originally had a non-block expression, but we can't fit on
+            // one line, so we'll insert a block.
+            rewrite_closure_with_block(context, body_shape, &prefix, body)
+        })
     }
+}
 
-    // The closure originally had a non-block expression, but we can't fit on
-    // one line, so we'll insert a block.
+fn rewrite_closure_with_block(
```

Comment:
Could you add a comment to explain the difference between `rewrite_closure_with_block` and `rewrite_closure_block` please?